test/e2e/gameserver_test.go

```diff
@@ -1147,3 +1147,50 @@ func TestPlayerConnectAndDisconnect(t *testing.T) {
 	assert.NoError(t, err)
 	assert.ElementsMatch(t, []string{"1", "3"}, gs.Status.Players.IDs)
 }
+
+func TestGracefulShutdown(t *testing.T) {
+	if !runtime.FeatureEnabled(runtime.FeatureSDKGracefulTermination) {
+		t.SkipNow()
+	}
+	t.Parallel()
+
+	log := e2eframework.TestLogger(t)
+	ctx := context.Background()
+	gs := framework.DefaultGameServer(framework.Namespace)
+	var minute int64 = 60
+	gs.Spec.Template.Spec.TerminationGracePeriodSeconds = &minute
+	readyGs, err := framework.CreateGameServerAndWaitUntilReady(t, framework.Namespace, gs)
+	if err != nil {
+		t.Fatalf("Could not get a GameServer ready: %v", err)
+	}
+	assert.Equal(t, readyGs.Status.State, agonesv1.GameServerStateReady)
+	gameservers := framework.AgonesClient.AgonesV1().GameServers(framework.Namespace)
+	err = gameservers.Delete(ctx, readyGs.ObjectMeta.Name, metav1.DeleteOptions{})
+	require.NoError(t, err)
+	log.Info("Created GameServer, waited 20 seconds")
+	time.Sleep(20 * time.Second)
+	log.WithField("gs", gs).Info("Checking GameServer")
+	gs, err = gameservers.Get(ctx, readyGs.ObjectMeta.Name, metav1.GetOptions{})
+	require.NoError(t, err)
+	assert.Equal(t, readyGs.ObjectMeta.Name, gs.ObjectMeta.Name)
+
+	// move it to shutdown
+	gsCopy := gs.DeepCopy()
+	gsCopy.Status.State = agonesv1.GameServerStateShutdown
+	_, err = gameservers.Update(ctx, gsCopy, metav1.UpdateOptions{})
```

Comment:
Why are you triggering the shutdown this way instead of sending `"EXIT"` to the game server so that it calls `sdk.Shutdown()` from within the game server?

The way I read the feature, the idea is that the sidecar waits to exit until the game server can clean up and call shutdown, so why not exercise that flow by having the game server call shutdown?

Reply:
So we actually can't at this point, because the `simple-game-server` has already exited! 

What we are really testing here is what happens when a `GameServer` `Pod` container gets a SIGTERM  signal, as if a Node was about to shut down under the Pod (or really any other reason a SIGTERM is fired). To get that to happen, we have to delete the `GameServer` as we do above.

At that point, the `simple-game-server` handles the SIGTERM as an exit signal, and exits - so we can't send it more UDP packets. The sdk-server though acts as we would want - hanging out until the `GameServer` moves to `Shutdown` state. which keeps the `GameServer` around. 

> The way I read the feature, the idea is that the sidecar waits to exit until the game server can clean up and call shutdown,

Right - but that was the thing I want to fix in this PR. The `sdk-server` shouldn't be to wait until the GameServer calls _just_ `sdk.Shutdown()` - it should exit as soon as _anything_ moved the `GameServer` to `Shutdown` - which includes scale down from a Fleet -- otherwise the sidecar is always going to wait the grace period on scale down, regardless of if the GameServer is in `Shutdown` state or not.

So running e2e test through the SDK doesn't actually do anything for our testing. We only really care that the GameServer is responding to a Shutdown state from _anywhere_. That's what actually matters.

It's a subtle difference that ultimately makes a big different to overall processing times for Fleets in general - does that make sense?

Reply:
> So we actually can't at this point, because the `simple-game-server` has already exited!

That makes sense. If we wanted to test that flow we could modify the simple game server to wait N seconds after receiving a SIGTERM and then call shutdown, but real game servers that aren't designed to handle SIGTERM may just immediately exit as the simple-game-server does today and we don't want the sidecar hanging around in those cases. 

I'm confused a bit about your statement about things other than the SDK moving the gameserver to the shutdown state though. My understanding based on https://agones.dev/site/docs/reference/gameserver/#gameserver-state-diagram is that the SDK is the way that a game server moves to shutdown. But it sounds like the fleet controller may also do this as part of scaling down the fleet (to remove replicas) - but presumably only on ready game servers and not on reserved or allocated ones. Are there other scenarios that you know of that would also cause the game server to be moved into the shutdown state through an external mechanism?

Reply:
> My understanding based on https://agones.dev/site/docs/reference/gameserver/#gameserver-state-diagram is that the SDK is the way that a game server moves to shutdown

This is true. That is how the SDK works.

> But it sounds like the fleet controller may also do this as part of scaling down the fleet (to remove replicas) - but presumably only on ready game servers and not on reserved or allocated ones.

See:
https://github.com/googleforgames/agones/blob/7492d86c2f55b3980050522f961f886507abbd96/pkg/gameserversets/controller.go#L498-L514

> Are there other scenarios that you know of that would also cause the game server to be moved into the shutdown state through an external mechanism?

🤔 There could be an argument that we the SDK server might want to exit early on `Unhealthy` or `Error` state, bit given that that state should be the exception rather than the norm, I'm happy to wait and see how things go with those states, before making that change.

I did a quick search of the codebase, but outside the SDK and the `GameServerSet` controller, those are the only places that Agones moves a GameServer to `Shutdown` (there was a good reason to do it on Fleet scale down, but now I forget what it was - possibly to have a shutdown event in the GameServer? - although with this change, it's now very important)

But in theory, someone could build a system external to Agones that moves a GameServer to Shutdown as an explicit message to Shutdown a GameServer - since it is an open API after all.

Reply:
Oh! I know why we had to move to Shutdown on scale down! Otherwise you could accidentally delete an Allocated `GameServer`.

If you looked at the cached version of the GameServer, it might say it was in Ready state, but you wouldn't know. Deletion doesn't look at generational locking, so you _have_ to change the `GameServer` object to ensure that the copy you are looking at is actually the current state of the `GameServer` on etcd.

See - I knew there was an important reason.

Reply:
This all makes sense, but makes me wonder if we should document this somewhere so it's discoverable (as I'm just learning about it now).